packages/ifpack/src/Ifpack_Hypre.cpp

```diff
@@ -91,36 +95,48 @@ Ifpack_Hypre::Ifpack_Hypre(Epetra_RowMatrix* A):
   for(int i = 0; i < Comm().NumProc()-1; i++){
     NiceRowMap_ = (NiceRowMap_ && iuppers[i]+1 == ilowers[i+1]);
   }
-  if(!NiceRowMap_){
-    ilower = (A_->NumGlobalRows() / Comm().NumProc())*Comm().MyPID();
-    iupper = (A_->NumGlobalRows() / Comm().NumProc())*(Comm().MyPID()+1)-1;
-    if(Comm().MyPID() == Comm().NumProc()-1){
-      iupper = A_-> NumGlobalRows()-1;
-    }
+  // create maps for the Hypre matrix
+  if (!NiceRowMap_) {
+    // Must create "nice" ColumnMap to map indices via LID.
+    //   Epetra_GID  --------->   LID   ----------> HYPRE_GID
+    //           via RowMap.LID()       via SimpleColumnMap.GID()
+    MySimpleMap_ = rcp(new Epetra_Map(-1, A_->RowMatrixRowMap().NumMyElements(),
+            0, Comm()));
+    Epetra_Import importer(A_->RowMatrixColMap(), A_->RowMatrixRowMap());
```

Comment:
By "trivial (null)" do you mean I could simply add a check here:

`if (A_->Exporter() != nullptr) { return SOME_MEANINGFUL_ERROR; } `

And is this condition necessary?  As in, could there be a more general case where the row map and the range map were permutations of each other, in which case the exporter was not NULL but the RowMap was not overlapping?  I assume this thing would have to be made with a user-provided RangeMap, and I don't know why people would do this...

Reply:
@ecoon wrote:
> By "trivial (null)" do you mean I could simply add a check here:

Yes

> And is this condition necessary?

No.  In fact, you could support both the "row Map is a one-to-one permutation of the range Map" and "row Map is not one to one" cases if you like.

  - "Row Map is one to one, but not same as range Map": Assuming that the domain and range Maps are the same, use the matrix's Export in reverse mode to permute the input vector to the row Map, call Hypre on it, then use the Export in forward mode to permute the output vector back to the range Map.
  - "Row Map is not one to one": Create a new empty matrix with a globally contiguous Map (preferably the range Map, or your new Map if the range Map is not globally contiguous).  Export with ADD CombineMode from the old matrix to the new matrix.  (If using the range Map, you may use the old matrix's Export object.  Otherwise, you must create a new Export object.)  Then, fill-complete the new matrix (supply the domain and range Map), and give that new matrix to Hypre.  Hypre will only know about the new matrix.

Most Ifpack preconditioners support neither of these cases, so it's up to you if you wish to support them.  Please do at least check for the condition, though, since it's cheap ("is Export nullptr?") and can save users puzzlement.

Reply:
@ecoon wrote:
> So `Epetra_CrsMatrix::Exporter()` exists, but `Epetra_RowMatrix::Exporter()` does not exist.

That's weird.  That may be an Epetra bug.  It's likely that users are giving you an `Epetra_CrsMatrix`, so you could just attempt a `dynamic_cast` to `Epetra_CrsMatrix`.  `Epetra_Map::SameAs` may have a nontrivial cost (worst case does an all-reduce), so it's better to try the `dynamic_cast` first, then go for `SameAs` if the cast fails.

Reply:
We are always using a CrsMatrix, but that isn't the interface...

I think the SameAs() is ok here -- this is in the Ifpack_Hypre constructor, which really really should only get called a handful of times by a typical application.  The only reason one would call it more than once per physical system solved is if you were doing not value changes (just call Compute()), not sparsity structure changes (just call Initialize()), but actually changing the maps.  

If you really want the dynamic cast, I'll do it, and it should get most cases right, but it seems like minor optimization at the cost of readability to me.

Reply:
@ecoon wrote:
> not sparsity structure changes (just call Initialize()), but actually changing the maps.

Epetra doesn't easily support sparsity structure changes without changing Maps too.




> If you really want the dynamic cast, I'll do it, and it should get most cases right, but it seems like minor optimization at the cost of readability to me.

It's up to your users.